MountRoulette addon

xurion opened the pull request:
Summons a mount at random based on your mount key items.

A function to blacklist certain mounts exists on another branch and isn't included in this PR because of the impact of https://github.com/Windower/Lua/issues/1916 (blacklists mounts unexpectedly across multiple characters).

xurion committed:
Added MountRoulette addon.

z16 commented on addons/MountRoulette/MountRoulette.lua:
```diff
@@ -0,0 +1,79 @@
+    end
+end)
+
+windower.register_event('addon command', function(command)
```
While this pattern by itself is nice for more complicated command handlers I feel it's unnecessarily complex here. I'd just make `//mr` run the roulette and remove the help text and all the boilerplate surrounding it, as well as the help message for it. I don't see a benefit to having this setup,

xurion replied:
No worries - this is the boilerplate left over from  removing the blacklist stuff. I'll cut this out as well.

z16 commented on addons/MountRoulette/MountRoulette.lua:
```diff
@@ -0,0 +1,79 @@
+    end
+end
+
+windower.register_event('load', function()
```
I would probably forego the `load` event here and just call the function directly. Its only purpose is if the code should execute after the full file has been loaded, but that does not seem to be necessary here.

z16 commented on addons/MountRoulette/MountRoulette.lua:
```diff
@@ -0,0 +1,79 @@
+            end
+        end
+    end
+end
```
I think this function needs to clear the `allowed_mounts` every time it's called. Otherwise if you log into a character that has all mounts, then log out and into a mule that has none, it will still have all mounts unlocked, no? Or am I missing something here?

xurion replied:
You're absolutely right, I'll clear it.

z16 commented on addons/MountRoulette/MountRoulette.lua:
```diff
@@ -0,0 +1,79 @@
+            local mount = possible_mounts[mount_index]
+
+            -- Add this to allowed mounts if it is not already there
+            if not allowed_mounts:contains(mount) then
```
If the `allowed_mounts` are cleared at the top of the function this check could be removed, right? That said, if that was made a set (`S{}`) you could remove it regardless, which is also what I would do, even if it's temporary and converted to a list later:
```lua
allowed_mounts = L(allowed_mounts_set)
```

z16 commented on addons/MountRoulette/MountRoulette.lua:
```diff
@@ -0,0 +1,79 @@
+
+windower.register_event('incoming chunk', function(id)
+    if id == 0x055 then --ki update
+        update_allowed_mounts()
```
Have you verified that this works once you get a new mount? I don't think it should, since the received KI packet will trigger the function, but at that point it hasn't reached the game yet, so `windower.ffxi.get_key_items()` would not know about it, I think. You may want to run it in a slightly delayed coroutine:
```lua
update_allowed_mounts:schedule(0.2)
```

xurion replied:
Ahh I did not know that - I've probably not ran into this issue due to getting new mounts in Upper Jeuno, then needing to go outside to summon a mount. Given this, do still think it's worth the schedule call?

xurion replied:
Thinking about it, EmpyPopTracker has this same pattern and works ok. Is it more of a race condition?

z16 replied:
Actually, now that I think about it this should work after all. We store our own copy of the key items, and `windower.ffxi.get_key_items` returns that, not the game's key item store, so it shouldn't matter. At least I think so...

If you have some mounts left to get you could test it by the function print out the mounts it has found, then you'll see if it's really there.

xurion replied:
I think an alt of mine has some mounts to turn in so I'll check that soon.

xurion replied:
Thankfully it seems to work fine:

![image](https://user-images.githubusercontent.com/606627/89741336-52560d00-da88-11ea-989c-321506a6a068.png)

xurion committed:
Switched update function to use sets instead.
Removed unnecessary load event.
Removed unnecessary command complexity.
PR comments.

xurion commented:
One thing I haven't considered is if someone uses this on a character with zero mounts. You'll get an error when it tries to concat the /mount with the mount name - should I cater for this?

z16 commented:
Ah why not, it's good form.

xurion committed:
Handle case where character has no mounts.

xurion commented:
Ok cool, I just threw a return statement in there rather than logging something.
